File: app/src/main/java/com/tunjid/rcswitchcontrol/fragments/ClientBleFragment.java
package com.tunjid.rcswitchcontrol.fragments;

import android.app.Activity;
import android.bluetooth.BluetoothDevice;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.snackbar.Snackbar;
import com.tunjid.androidbootstrap.material.animator.FabExtensionAnimator;
import com.tunjid.androidbootstrap.recyclerview.ListManager;
import com.tunjid.androidbootstrap.recyclerview.ListManagerBuilder;
import com.tunjid.androidbootstrap.recyclerview.ListPlaceholder;
import com.tunjid.androidbootstrap.recyclerview.SwipeDragOptionsBuilder;
import com.tunjid.androidbootstrap.view.animator.ViewHider;
import com.tunjid.rcswitchcontrol.R;
import com.tunjid.rcswitchcontrol.abstractclasses.BaseFragment;
import com.tunjid.rcswitchcontrol.activities.MainActivity;
import com.tunjid.rcswitchcontrol.adapters.RemoteSwitchAdapter;
import com.tunjid.rcswitchcontrol.dialogfragments.NameServiceDialogFragment;
import com.tunjid.rcswitchcontrol.dialogfragments.RenameSwitchDialogFragment;
import com.tunjid.rcswitchcontrol.model.RcSwitch;
import com.tunjid.rcswitchcontrol.services.ClientBleService;
import com.tunjid.rcswitchcontrol.services.ServerNsdService;
import com.tunjid.rcswitchcontrol.utils.DeletionHandler;
import com.tunjid.rcswitchcontrol.viewmodels.BleClientViewModel;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;
import androidx.transition.AutoTransition;
import androidx.transition.TransitionManager;

import static com.tunjid.rcswitchcontrol.App.isServiceRunning;
import static com.tunjid.rcswitchcontrol.services.ClientBleService.ACTION_CONTROL;
import static com.tunjid.rcswitchcontrol.services.ClientBleService.ACTION_SNIFFER;
import static com.tunjid.rcswitchcontrol.services.ClientBleService.BLUETOOTH_DEVICE;
import static java.util.Objects.requireNonNull;

public class ClientBleFragment extends BaseFragment
        implements
        RemoteSwitchAdapter.SwitchListener,
        RenameSwitchDialogFragment.SwitchNameListener,
        NameServiceDialogFragment.ServiceNameListener {

    private int lastOffSet;
    private boolean isDeleting;

    private View progressBar;
    private TextView connectionStatus;
    private ListManager<RemoteSwitchAdapter.ViewHolder, ListPlaceholder> listManager;

    private ViewHider viewHider;
    private BleClientViewModel viewModel;

    public static ClientBleFragment newInstance(BluetoothDevice bluetoothDevice) {
        ClientBleFragment fragment = new ClientBleFragment();
        Bundle args = new Bundle();
        args.putParcelable(BLUETOOTH_DEVICE, bluetoothDevice);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        viewModel = ViewModelProviders.of(requireActivity()).get(BleClientViewModel.class);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        View root = inflater.inflate(R.layout.fragment_ble_client, container, false);
        AppBarLayout appBarLayout = root.findViewById(R.id.app_bar_layout);

        progressBar = root.findViewById(R.id.progress_bar);
        connectionStatus = root.findViewById(R.id.connection_status);

        listManager = new ListManagerBuilder<RemoteSwitchAdapter.ViewHolder, ListPlaceholder>()
                .withRecyclerView(root.findViewById(R.id.switch_list))
                .withAdapter(new RemoteSwitchAdapter(this, viewModel.getSwitches()))
                .withLinearLayoutManager()
                .addScrollListener((dx, dy) -> {
                    if (dy == 0) return;
                    if (dy > 0) viewHider.hide();
                    else viewHider.show();
                })
                .withSwipeDragOptions(new SwipeDragOptionsBuilder<RemoteSwitchAdapter.ViewHolder>()
                        .setMovementFlagsFunction(holder -> getSwipeDirection())
                        .setSwipeConsumer(((viewHolder, direction) -> onDelete(viewHolder)))
                        .build())
                .build();

        viewHider = ViewHider.of(root.findViewById(R.id.button_container))
                .setDuration(ViewHider.BOTTOM).build();

        appBarLayout.addOnOffsetChangedListener((appBarLayout1, verticalOffset) -> {
            if (verticalOffset == 0) return;
            if (verticalOffset > lastOffSet) viewHider.hide();
            else viewHider.show();

            lastOffSet = verticalOffset;
        });

        return root;
    }

    @Override
    public void onResume() {
        super.onResume();
        getToolBar().setTitle(R.string.switches);

        BluetoothDevice device = requireNonNull(getArguments()).getParcelable(BLUETOOTH_DEVICE);
        disposables.add(viewModel.listenBle(device).subscribe(this::onReceive, Throwable::printStackTrace));
        disposables.add(viewModel.connectionState().subscribe(this::onConnectionStateChanged, Throwable::printStackTrace));
        disposables.add(viewModel.listenServer().subscribe(connected -> requireActivity().invalidateOptionsMenu(), Throwable::printStackTrace));
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_ble_client, menu);

        menu.findItem(R.id.menu_start_nsd).setVisible(!isServiceRunning(ServerNsdService.class));
        menu.findItem(R.id.menu_restart_nsd).setVisible(isServiceRunning(ServerNsdService.class));

        if (viewModel.isBleBound()) {
            menu.findItem(R.id.menu_connect).setVisible(!viewModel.isBleConnected());
            menu.findItem(R.id.menu_disconnect).setVisible(viewModel.isBleConnected());
        }

        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (!viewModel.isBleBound()) return super.onOptionsItemSelected(item);

        switch (item.getItemId()) {
            case R.id.menu_connect:
                viewModel.reconnectBluetooth();
                return true;
            case R.id.menu_disconnect:
                viewModel.disconnectBluetooth();
                return true;
            case R.id.menu_start_nsd:
                NameServiceDialogFragment.newInstance().show(getChildFragmentManager(), "");
                break;
            case R.id.menu_restart_nsd:
                viewModel.restartServer();
                break;
            case R.id.menu_forget:
                viewModel.forgetBluetoothDevice();

                Activity activity = requireActivity();
                activity.finish();

                startActivity(new Intent(activity, MainActivity.class)
                        .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDestroyView() {
        listManager.clear();
        listManager = null;
        progressBar = null;
        connectionStatus = null;

        super.onDestroyView();
    }

    @Override protected boolean showsFab() {
        return true;
    }

    @Override protected FabExtensionAnimator.GlyphState getFabState() {
        return viewModel.getFabState();
    }

    @Override protected View.OnClickListener getFabClickListener() {
        return view -> {
            toggleProgress(true);
            viewModel.sniffRcSwitch();
        };
    }

    @Override
    public void onLongClicked(RcSwitch rcSwitch) {
        RenameSwitchDialogFragment.newInstance(rcSwitch).show(getChildFragmentManager(), "");
    }

    @Override
    public void onSwitchToggled(RcSwitch rcSwitch, boolean state) {
        viewModel.toggleSwitch(rcSwitch, state);
    }

    @Override
    public void onSwitchRenamed(RcSwitch rcSwitch) {
        listManager.notifyItemChanged(viewModel.onSwitchUpdated(rcSwitch));
    }

    @Override
    public void onServiceNamed(String name) {
        viewModel.nameServer(name);
    }

    private void onConnectionStateChanged(String status) {
        requireActivity().invalidateOptionsMenu();
        connectionStatus.setText(status);
    }

    private void toggleProgress(boolean show) {
        TransitionManager.beginDelayedTransition((ViewGroup) progressBar.getParent(), new AutoTransition());
        progressBar.setVisibility(show ? View.VISIBLE : View.INVISIBLE);
        togglePersistentUi();
    }

    private void onReceive(Intent intent) {
        String action = intent.getAction();
        if (action == null) return;

        switch (action) {
            case ACTION_CONTROL:
                byte[] rawData = intent.getByteArrayExtra(ClientBleService.DATA_AVAILABLE_CONTROL);
                toggleProgress(rawData[0] == 0);
                break;
            case ACTION_SNIFFER: {
                listManager.notifyDataSetChanged();
                toggleProgress(false);
                break;
            }
        }
    }

    private int getSwipeDirection() {
        return isDeleting ? 0 : ItemTouchHelper.SimpleCallback.makeMovementFlags(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT);
    }

    private void onDelete(RecyclerView.ViewHolder viewHolder) {
        if (isDeleting) return;
        isDeleting = true;

        View root = getView();

        if (root == null) return;
        int position = viewHolder.getAdapterPosition();
        List<RcSwitch> switches = viewModel.getSwitches();

        DeletionHandler<RcSwitch> deletionHandler = new DeletionHandler<>(position, () -> {
            isDeleting = false;
            RcSwitch.saveSwitches(switches);
        });

        deletionHandler.push(switches.get(position));

        Snackbar.make(root, R.string.deleted_switch, Snackbar.LENGTH_LONG)
                .addCallback(deletionHandler)
                .setAction(R.string.undo, view -> {
                    if (!deletionHandler.hasItems()) return;

                    int deletedAt = deletionHandler.getDeletedPosition();
                    switches.add(deletedAt, deletionHandler.pop());
                    listManager.notifyItemInserted(deletedAt);

                    isDeleting = false;
                })
                .show();
    }
}

File: app/src/main/java/com/tunjid/rcswitchcontrol/fragments/ClientBleFragment.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.tunjid.rcswitchcontrol.fragments

import android.bluetooth.BluetoothDevice
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.TextView
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.ItemTouchHelper.Callback.makeMovementFlags
import androidx.recyclerview.widget.RecyclerView
import androidx.transition.AutoTransition
import androidx.transition.TransitionManager
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.snackbar.Snackbar
import com.tunjid.androidbootstrap.material.animator.FabExtensionAnimator
import com.tunjid.androidbootstrap.recyclerview.ListManager
import com.tunjid.androidbootstrap.recyclerview.ListManagerBuilder
import com.tunjid.androidbootstrap.recyclerview.ListPlaceholder
import com.tunjid.androidbootstrap.recyclerview.SwipeDragOptionsBuilder
import com.tunjid.rcswitchcontrol.App.Companion.isServiceRunning
import com.tunjid.rcswitchcontrol.R
import com.tunjid.rcswitchcontrol.abstractclasses.BaseFragment
import com.tunjid.rcswitchcontrol.activities.MainActivity
import com.tunjid.rcswitchcontrol.adapters.DeviceAdapter
import com.tunjid.rcswitchcontrol.data.Device
import com.tunjid.rcswitchcontrol.data.RfSwitch
import com.tunjid.rcswitchcontrol.data.ZigBeeDevice
import com.tunjid.rcswitchcontrol.dialogfragments.NameServiceDialogFragment
import com.tunjid.rcswitchcontrol.dialogfragments.RenameSwitchDialogFragment
import com.tunjid.rcswitchcontrol.services.ClientBleService.Companion.BLUETOOTH_DEVICE
import com.tunjid.rcswitchcontrol.services.ServerNsdService
import com.tunjid.rcswitchcontrol.utils.DeletionHandler
import com.tunjid.rcswitchcontrol.utils.SpanCountCalculator
import com.tunjid.rcswitchcontrol.viewmodels.BleClientViewModel

open class ClientBleFragment : BaseFragment(),
        DeviceAdapter.Listener,
        RenameSwitchDialogFragment.SwitchNameListener,
        NameServiceDialogFragment.ServiceNameListener {

    private var lastOffSet: Int = 0
    private var isDeleting: Boolean = false

    private lateinit var progressBar: View
    private lateinit var connectionStatus: TextView
    private lateinit var listManager: ListManager<ViewHolder, ListPlaceholder<*>>

    private lateinit var viewModel: BleClientViewModel

    override val fabState: FabExtensionAnimator.GlyphState
        get() = viewModel.fabState

    override val fabClickListener: View.OnClickListener
        get() = View.OnClickListener {
            toggleProgress(true)
            viewModel.sniffRcSwitch()
        }

    private val swipeDirection: Int
        get() = if (isDeleting) 0 else makeMovementFlags(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT)

    override fun onAttach(context: Context) {
        super.onAttach(context)
        viewModel = ViewModelProviders.of(requireActivity()).get(BleClientViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val root = inflater.inflate(R.layout.fragment_ble_client, container, false)
        val appBarLayout = root.findViewById<AppBarLayout>(R.id.app_bar_layout)

        progressBar = root.findViewById(R.id.progress_bar)
        connectionStatus = root.findViewById(R.id.connection_status)

        listManager = ListManagerBuilder<ViewHolder, ListPlaceholder<*>>()
                .withRecyclerView(root.findViewById(R.id.switch_list))
                .withAdapter(DeviceAdapter(this, viewModel.switches))
                .withGridLayoutManager(SpanCountCalculator.spanCount)
                .addScrollListener { _, dy -> if (dy != 0) toggleFab(dy < 0) }
                .withSwipeDragOptions(SwipeDragOptionsBuilder<ViewHolder>()
                        .setMovementFlagsFunction { swipeDirection }
                        .setSwipeConsumer { viewHolder, _ -> onDelete(viewHolder) }
                        .build())
                .build()

        appBarLayout.addOnOffsetChangedListener(object : AppBarLayout.OnOffsetChangedListener {
            override fun onOffsetChanged(p0: AppBarLayout?, verticalOffset: Int) {
                if (verticalOffset == 0) return
                toggleFab(verticalOffset < lastOffSet)
                lastOffSet = verticalOffset
            }
        })

        return root
    }

    override fun onResume() {
        super.onResume()
        toolBar.setTitle(R.string.switches)

        disposables.add(viewModel.connectionState().subscribe(this::onConnectionStateChanged, Throwable::printStackTrace))
        disposables.add(viewModel.listenServer().subscribe({ requireActivity().invalidateOptionsMenu() }, Throwable::printStackTrace))
        arguments!!.getParcelable<BluetoothDevice>(BLUETOOTH_DEVICE)?.let {
            disposables.add(viewModel.listenBle(it).subscribe(this::toggleProgress, Throwable::printStackTrace))
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_ble_client, menu)

        val bleConnected = viewModel.isBleConnected
        val serviceRunning = isServiceRunning(ServerNsdService::class.java)

        menu.findItem(R.id.menu_refresh).isVisible = bleConnected
        menu.findItem(R.id.menu_connect).isVisible = !bleConnected
        menu.findItem(R.id.menu_disconnect).isVisible = bleConnected
        menu.findItem(R.id.menu_start_nsd).isVisible = !serviceRunning
        menu.findItem(R.id.menu_restart_nsd).isVisible = serviceRunning

        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (!viewModel.isBleBound) return super.onOptionsItemSelected(item)

        when (item.itemId) {
            R.id.menu_refresh -> {
                viewModel.refreshSwitches()
                listManager.notifyDataSetChanged()
                return true
            }
            R.id.menu_connect -> {
                viewModel.reconnectBluetooth()
                return true
            }
            R.id.menu_disconnect -> {
                viewModel.disconnectBluetooth()
                return true
            }
            R.id.menu_start_nsd -> NameServiceDialogFragment.newInstance().show(childFragmentManager, "")
            R.id.menu_restart_nsd -> viewModel.restartServer()
            R.id.menu_forget -> {
                viewModel.forgetBluetoothDevice()

                val activity = requireActivity()
                activity.finish()

                startActivity(Intent(activity, MainActivity::class.java)
                        .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onDestroyView() {
        listManager.clear()
        super.onDestroyView()
    }

    override fun showsFab(): Boolean = true

    override fun onLongClicked(rfSwitch: RfSwitch) =
            RenameSwitchDialogFragment.newInstance(rfSwitch).show(childFragmentManager, "")

    override fun onSwitchToggled(rfSwitch: RfSwitch, state: Boolean) =
            viewModel.toggleSwitch(rfSwitch, state)

    override fun onSwitchRenamed(rfSwitch: RfSwitch) =
            listManager.notifyItemChanged(viewModel.onSwitchUpdated(rfSwitch))

    override fun onLongClicked(device: ZigBeeDevice) {
    }

    override fun onSwitchToggled(device: ZigBeeDevice, state: Boolean) {
    }

    override fun rediscover(device: ZigBeeDevice) {
    }

    override fun color(device: ZigBeeDevice) {
    }

    override fun level(device: ZigBeeDevice, level: Float) {
    }

    override fun onServiceNamed(name: String) = viewModel.nameServer(name)

    private fun onConnectionStateChanged(status: String) {
        requireActivity().invalidateOptionsMenu()
        connectionStatus.text = status
    }

    private fun toggleProgress(show: Boolean) {
        TransitionManager.beginDelayedTransition(progressBar.parent as ViewGroup, AutoTransition())
        progressBar.visibility = if (show) View.VISIBLE else View.INVISIBLE
        togglePersistentUi()
    }

    private fun onDelete(viewHolder: RecyclerView.ViewHolder) {
        if (isDeleting) return
        isDeleting = true

        val root = view ?: return

        val position = viewHolder.adapterPosition
        val switches = viewModel.switches

        val deletionHandler = DeletionHandler<Device>(position) {
            isDeleting = false
            viewModel.saveSwitches()
        }

        deletionHandler.push(switches[position])
        switches.removeAt(position)
        listManager.notifyItemRemoved(position)

        Snackbar.make(root, R.string.deleted_switch, Snackbar.LENGTH_LONG)
                .addCallback(deletionHandler)
                .setAction(R.string.undo) {
                    if (deletionHandler.hasItems()) {
                        val deletedAt = deletionHandler.deletedPosition
                        switches.add(deletedAt, deletionHandler.pop())
                        listManager.notifyItemInserted(deletedAt)

                        isDeleting = false
                    }
                }
                .show()
    }

    companion object {

        fun newInstance(bluetoothDevice: BluetoothDevice): ClientBleFragment {
            val fragment = ClientBleFragment()
            val args = Bundle()
            args.putParcelable(BLUETOOTH_DEVICE, bluetoothDevice)
            fragment.arguments = args
            return fragment
        }
    }
}
